Sprint 17

boyegorka opened the pull request:
Выполнены задания по 17 спринту

boyegorka committed:
Sprint-17
Реализована локализация
Реализованы пункты у трекера "Редактировать" и "Удалить"

boyegorka committed:
Sprint-17
Добавлена AppMetrica с помощью podfile

boyegorka committed:
Sprint-17
Добавление pod файлов

boyegorka committed:
Sprint-17
Скриншот тесты добавлены в приложение

boyegorka committed:
Sprint-17
Выполнены задания спринта

So1ovyov commented on Tracker/Views/Cells/CollectionCell.swift:
```diff
@@ -34,9 +34,16 @@ final class CollectionCell: UITableViewCell {
+    var selectedIndex: Int? {
+        didSet {
+            guard let selectedIndex else { return }
+            collectionView.selectItem(at: IndexPath.init(row: selectedIndex, section: 0), animated: true, scrollPosition: .centeredVertically)
```
Чтобы легче читалось, давай расположим так:
```
            collectionView.selectItem(
                at: IndexPath.init(row: selectedIndex, section: 0),
                animated: true,
                scrollPosition: .centeredVertically
            )
```

So1ovyov commented on Tracker/ViewControllers/Trackers/TrackersViewController.swift:
```diff
@@ -237,9 +273,60 @@ extension TrackersViewController: UICollectionViewDelegate {
+        
+        let context = UIContextMenuConfiguration(identifier: nil, previewProvider: nil) { (action) -> UIMenu? in
+            
+            let pin = UIAction(title: tracker.isPinned ? "unpin".localized : "pin".localized,
```
Давай придумаем какой-нибудь ActionFactory, в которой вынесем всю логику по созданию UIAction и тем самым разгрузим метод collectionView.

So1ovyov commented on Tracker/ViewControllers/TabBarController.swift:
```diff
@@ -23,13 +26,20 @@ final class TabBarController: UITabBarController {
         let trackers = UINavigationController(rootViewController: vc)
-        trackers.tabBarItem = UITabBarItem(title: "Трекеры", image: UIImage(systemName: "record.circle.fill"), selectedImage: nil)
+        trackers.tabBarItem = UITabBarItem(title: "trackers".localized, image: UIImage(systemName: "record.circle.fill"), selectedImage: nil)
+        analytics.report(event: "click", params: ["item":"open_trackers_screen"])
```
Не корректное применение аналитики, тут нет никакого события Click. 
По факту это событие произойдет когда у тебя запуститься приложение и загрузится tabBarController.
Отсюда нужно удалить это событие аналитик и добавить в тот момент, где обрабатывается нажатие на таб бар.

boyegorka replied:
Про какие фильтры идёт речь в задании про события в Яндекс метрике?

boyegorka replied:
> Не корректное применение аналитики, тут нет никакого события Click. По факту это событие произойдет когда у тебя запуститься приложение и загрузится tabBarController. Отсюда нужно удалить это событие аналитик и добавить в тот момент, где обрабатывается нажатие на таб бар.

Про какие фильтры идёт речь в задании по Яндекс метрике?

boyegorka replied:
> Не корректное применение аналитики, тут нет никакого события Click. По факту это событие произойдет когда у тебя запуститься приложение и загрузится tabBarController. Отсюда нужно удалить это событие аналитик и добавить в тот момент, где обрабатывается нажатие на таб бар.

"❌ track для тапа на кнопку о выполнении трека" - оно находится в ячейке

boyegorka replied:
> Не корректное применение аналитики, тут нет никакого события Click. По факту это событие произойдет когда у тебя запуститься приложение и загрузится tabBarController. Отсюда нужно удалить это событие аналитик и добавить в тот момент, где обрабатывается нажатие на таб бар.

Посмотрел, что за пункт с фильтром, только не говори, что мне придётся реализовывать фильтры, только чтобы я прошёл проверку на метрику 😢

So1ovyov replied:
Привет @boyegorka ! Просто нужно чтобы было как в ТЗ. Определенные события с определенным наименованием метрики
<img width="456" alt="Снимок экрана 2023-10-05 в 08 47 15" src="https://github.com/boyegorka/Tracker/assets/68296988/e3fa8045-6feb-486d-b345-24c089ada2b3">

So1ovyov replied:
Сделай так чтобы было событие click с наименованием item как на скриншоте

boyegorka replied:
> Сделай так чтобы было событие click с наименованием item как на скриншоте
Даже неважно, будет оно работать или нет?

So1ovyov replied:
@boyegorka кнопка Фильтры, ее функционал и событие метрики также должно быть в проекте. Без этого работа не зачтется, это же тоже часть ТЗ.

boyegorka replied:
> @boyegorka кнопка Фильтры, ее функционал и событие метрики также должно быть в проекте. Без этого работа не зачтется, это же тоже часть ТЗ.

На протяжение всего проекта не было написано ничего про фильтры(

So1ovyov replied:
@boyegorka это очень странно, но в макете они есть, и даже событие на эту кнопку есть

boyegorka replied:
> @boyegorka это очень странно, но в макете они есть, и даже событие на эту кнопку есть

Как я могу реализовать фильтры? Мне же нужно чтобы они всегда были применены, это скорее всего должен быть дополнительный предикат, который будет в сервисе. Но как мне применить нужный предикат? Я могу сделать enum в filtersViewContreller и передавать что-то, но что? Имя текстом или номер? По идее я же могу как-то сделать в сервисе тип того enum, который будет во вью? Пока не понимаю через что его делать.

So1ovyov commented on Tracker/ViewControllers/Statistics/StatisticsViewController.swift:
```diff
@@ -7,53 +7,181 @@
+        presenter?.getAllTrackersStat()
+    }
+    
+    private func statsAreZero() -> Bool {
```
Есть ли необходимость выносить логику в этот метод и метод noStats ниже. Так как они не переиспользуются можно условия в этих методах вынести в метод setupEmptyScreen, пример распишу в метод setupEmptyScreen

So1ovyov commented on Tracker/ViewControllers/Statistics/StatisticsViewController.swift:
```diff
@@ -7,53 +7,181 @@
-        view.backgroundColor = .ypWhite
-        setupNavigationBar()
+    // MARK: - Public Methods
+    func setupEmptyScreen() {
```
Можем сократить код, и удалить 2 метода, например сделав что-то подобное:
```
    func setupEmptyScreen() {
        let emptyStats = StatisticsType.allCases.count <= 0
        let zeroStats = presenter?.completedTrackerCount == 0
        let isEmpty = emptyStats && zeroStats
        emptyScreenView.isHidden = !isEmpty
        statisticsTableView.isHidden = isEmpty
    }
```

So1ovyov commented on Tracker/ViewControllers/NewHabit/NewTrackerViewController.swift:
```diff
@@ -327,9 +359,16 @@ extension NewTrackerViewController: UITableViewDelegate {
+        switch rowsForSection(sections[indexPath.section])[indexPath.row] {
+        case .daysCounter:
+            switch presenter?.state {
+            case .new:
```
Значение .new и nil имеют одинаковый return может имеет смысл их объеденить или использовать default ?
```
        case .daysCounter:
            switch presenter?.state {
            case .new, nil:
                return .zero
            case .edit:
                return 50
            }
```
или так
```
            switch presenter?.state {
            case .edit:
                return 50
            default:
                return 0
            }
```

So1ovyov commented on Tracker/ViewControllers/NewHabit/NewTrackerViewController.swift:
```diff
@@ -15,19 +15,22 @@ final class NewTrackerViewController: UIViewController, NewTrackerViewController
     
     // MARK: - Enums
     enum Constant {
```
Круто что выносишь в Constant ! но если это небходимо использовать в рамках NewTrackerViewController.swift давай сделаем этот enum приватным

So1ovyov commented on Tracker/ViewControllers/Category/NewCategory/NewCategoryViewController.swift:
```diff
@@ -7,17 +7,21 @@
 
 import UIKit
 
+protocol CNewCategoryViewControllerProtocol: AnyObject {
```
Кажется в нейминг лишнее попало `NewCategoryViewControllerProtocol` (без C)

So1ovyov commented on Tracker/ViewControllers/AlertPresenter.swift:
```diff
@@ -0,0 +1,30 @@
+        let action = UIAlertAction(title: result.buttonText, style: .destructive) {_ in
+            result.completion()
+        }
+        let cancel = UIAlertAction(title: "Отменить", style: .cancel) { _ in
```
Может сделаем более короткую запись без раскрытия completion
```
(UIAlertAction(title: "Нет", style: .cancel, handler: nil))
```

boyegorka committed:
Sprint-17
Исправил замечания куратора

p.s. Про фильтры не было сказано ни в одном спринте, нужно ли их реализовывать, только чтобы реализовать метрику?

So1ovyov commented:
@boyegorka по поводу кнопки Фильтры, давай поступим следующим образом. Просто нарисуй эту кнопку и по клику на нее пусть запускается событие аналитики. Таким образом будет выполнено все в рамках ТЗ.
А по поводу самих событий сделай нейминг item-ов как на скриншоте выше.

boyegorka commented:
> @boyegorka по поводу кнопки Фильтры, давай поступим следующим образом. Просто нарисуй эту кнопку и по клику на нее пусть запускается событие аналитики. Таким образом будет выполнено все в рамках ТЗ. А по поводу самих событий сделай нейминг item-ов как на скриншоте выше.

Хорошо, сейчас сделаю

boyegorka committed:
Sprint-17
Исправлены замечания куратора

boyegorka commented:
> @boyegorka по поводу кнопки Фильтры, давай поступим следующим образом. Просто нарисуй эту кнопку и по клику на нее пусть запускается событие аналитики. Таким образом будет выполнено все в рамках ТЗ. А по поводу самих событий сделай нейминг item-ов как на скриншоте выше.

А сейчас?

So1ovyov commented:
@boyegorka почисти проект от закомментированного кода и возвращай работу мне на проверку, как будешь уверен что все поправил и все готово

boyegorka committed:
Sprint-17
Исправлены замечания куратора

boyegorka commented:
> @boyegorka почисти проект от закомментированного кода и возвращай работу мне на проверку, как будешь уверен что все поправил и все готово

Что-то я уже вообще ни в чём не уверен


> @boyegorka почисти проект от закомментированного кода и возвращай работу мне на проверку, как будешь уверен что все поправил и все готово

Ну, почистил проект, можно считать, что всё поправлено? Там последняя попытка на сдачу просто 😅

So1ovyov commented:
> > @boyegorka почисти проект от закомментированного кода и возвращай работу мне на проверку, как будешь уверен что все поправил и все готово
> 
> Что-то я уже вообще ни в чём не уверен
> 
> > @boyegorka почисти проект от закомментированного кода и возвращай работу мне на проверку, как будешь уверен что все поправил и все готово
> 
> Ну, почистил проект, можно считать, что всё поправлено? Там последняя попытка на сдачу просто 😅

Не переживай, если все проверил, возвращай на проверку.

boyegorka commented:
> > > @boyegorka почисти проект от закомментированного кода и возвращай работу мне на проверку, как будешь уверен что все поправил и все готово
> > 
> > 
> > Что-то я уже вообще ни в чём не уверен
> > > @boyegorka почисти проект от закомментированного кода и возвращай работу мне на проверку, как будешь уверен что все поправил и все готово
> > 
> > 
> > Ну, почистил проект, можно считать, что всё поправлено? Там последняя попытка на сдачу просто 😅
> 
> Не переживай, если все проверил, возвращай на проверку.

Отправил 🙄

So1ovyov commented on Tracker/Services/TrackerService.swift:
```diff
@@ -60,6 +66,20 @@ final class TrackerService: NSObject {
+
+        fetchRequest.predicate = NSPredicate(format: "%K == true", #keyPath(TrackerCoreData.isPinned))
+
+        let fetchedController = NSFetchedResultsController(fetchRequest: fetchRequest, managedObjectContext: persistentContainer.viewContext, sectionNameKeyPath: nil, cacheName: nil)
```
Давай разделим строки чтобы легче читалось:
```
        let fetchedController = NSFetchedResultsController(
            fetchRequest: fetchRequest,
            managedObjectContext: persistentContainer.viewContext,
            sectionNameKeyPath: nil,
            cacheName: nil
        )
```

So1ovyov commented on Tracker/Services/TrackerService.swift:
```diff
@@ -72,10 +92,12 @@ final class TrackerService: NSObject {
+            fetchRequest.predicate = NSCompoundPredicate(type: .and, subpredicates: [datePredicate, pinnedPredicate])
         }
 
         let fetchedController = NSFetchedResultsController(fetchRequest: fetchRequest, managedObjectContext: persistentContainer.viewContext, sectionNameKeyPath: #keyPath(TrackerCoreData.category.name), cacheName: nil)
```
И здесь давай также сделаем:
```
        let fetchedController = NSFetchedResultsController(
            fetchRequest: fetchRequest,
            managedObjectContext: persistentContainer.viewContext,
            sectionNameKeyPath: #keyPath(TrackerCoreData.category.name),
            cacheName: nil
        )
```

So1ovyov commented on Tracker/ViewControllers/Filters/FiltersViewController.swift:
```diff
@@ -0,0 +1,104 @@
+
+final class FiltersViewController: UIViewController, FiltersViewControllerProtocol {
+    
+    private enum Contstants {
```
Круто что выносишь все захардкоженные значения в private enum

So1ovyov commented on Tracker/ViewControllers/Statistics/StatisticsViewController.swift:
```diff
@@ -7,53 +7,180 @@
+
+extension StatisticsViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return 100
```
Если захочешь можно и эти захардкоженные значения вынести в enum:
```
static let tableViewHeight: CGFloat = 100
```

So1ovyov commented on Tracker/ViewControllers/Trackers/TrackersPresenter.swift:
```diff
@@ -63,22 +70,49 @@ final class TrackersPresenter: TrackersPresenterProtocol {
+        do {
+            try service.saveTracker(tracker, at: category)
+        } catch {
+            print(error)
```
В случае ошибки будет просто Print, было бы круто UI-но обработать жэту ошибку, чтобы пользователь понимал что происходит.

So1ovyov commented on Tracker/ViewControllers/Trackers/TrackersPresenter.swift:
```diff
@@ -63,22 +70,49 @@ final class TrackersPresenter: TrackersPresenterProtocol {
+        do {
+            try service.deleteTracker(tracker)
+        } catch {
+            print(error)
```
Также здесь и ниже можно подумать над универсальным alert-ом

So1ovyov commented on Tracker/ViewControllers/Trackers/TrackersViewController.swift:
```diff
@@ -237,9 +308,62 @@ extension TrackersViewController: UICollectionViewDelegate {
+        
+        let context = UIContextMenuConfiguration(identifier: nil, previewProvider: nil) { (action) -> UIMenu? in
+            
+            let pin = UIAction(title: tracker.isPinned ? "unpin".localized : "pin".localized,
```
Давай придумаем какой-нибудь ActionFactory, в которой вынесем всю логику по созданию UIAction и тем самым разгрузим метод collectionView.

boyegorka commented:
@So1ovyov Спасибо)
